File: Python/Two Furthest Houses With Different Colors.py
from typing import List


class Solution:
    """
    Time:   O(n)
    Memory: O(1)
    """

    def maxDistance(self, colors: List[int]) -> int:
        max_dist = 0

        for i, x in enumerate(colors):
            if x != colors[0]:
                max_dist = max(max_dist, i)
            if x != colors[-1]:
                max_dist = max(max_dist, len(colors) - 1 - i)

        return max_dist

File: Java/Two Furthest Houses With Different Colors.java
class Solution {

    /**
     * Time:   O(n)
     * Memory: O(1)
     */
    public int maxDistance(int[] colors) {
        int maxDist = 0, n = colors.length;

        for (int i = 0; i < n; ++i) {
            if (colors[i] != colors[0])
                maxDist = Math.max(maxDist, i);
            if (colors[i] != colors[n - 1])
                maxDist = Math.max(maxDist, n - i - 1);
        }

        return maxDist;
    }
}
